In [2]:
import numpy as np
import pandas as pd
import plotly.express as px
from sklearn.preprocessing import OrdinalEncoder
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import RFE

In [3]:
fs = pd.read_csv(r"..\..\data\pp\preprocessed_properties.csv")
fs

,property_type,society,sector,price_in_cr,area,bedRoom,bathroom,balcony,floornum_category,total_floor,...,ov_park,ov_pool,ov_club,ov_main_road,ov_sea,ov_others,dist_to_cyber_city,dist_to_golf_road,dist_to_airport,dist_to_manesar
0,Flat,smart world gems,sector 89,1.400,1196.0,2,2,2.0,Low-rise,4.0,...,0,0,0,0,0,1,17.442210,17.194023,22.471484,7.577067
1,Independent Builder Floor,other,suncity,2.500,2250.0,3,3,2.0,Low-rise,3.0,...,1,0,0,0,0,0,7.326951,1.580030,12.263103,21.726818
2,Flat,signature global city 92,sector 92,1.300,1091.0,3,2,3.0,Low-rise,4.0,...,1,1,1,1,1,1,21.144571,20.962383,25.996431,6.862454
3,Flat,m3m antalya hills,sector 79,2.250,2146.0,3,3,1.0,Low-rise,4.0,...,0,0,0,0,0,0,19.988459,17.314325,25.826044,3.975275
4,Flat,signature global park,sohna,0.815,845.0,2,2,3.0,Low-rise,4.0,...,1,1,1,1,0,1,23.125032,17.695935,28.980804,15.677376
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
39060,Flat,bestech park view sanskruti,sector 92,2.350,2120.0,3,4,4.0,Mid-rise,19.0,...,1,1,1,1,1,1,20.342766,19.985085,25.313540,6.317128
39061,Flat,tulip purple,sector 69,2.820,3308.0,4,5,2.0,Mid-rise,12.0,...,1,1,1,0,0,0,12.631193,9.416544,18.643203,11.556345
39062,Independent Builder Floor,other,sector 47,2.850,2367.0,3,3,2.0,Low-rise,3.0,...,1,0,0,0,0,0,8.509836,5.676428,14.543720,15.337482
39063,Flat,ashiana mulberry,sohna,1.220,1465.0,2,2,4.0,Mid-rise,17.0,...,0,0,0,1,0,0,24.966606,19.444161,30.782352,17.003169


In [4]:
fs.isnull().sum()

property_type              0
society                    0
sector                     0
price_in_cr                0
area                       0
bedRoom                    0
bathroom                   0
balcony                    0
floornum_category          0
total_floor                0
facing                     0
furnishing                 0
age_possession_category    0
is_corner                  0
covered_parking            0
open_parking               0
total_parking              0
has_ac                     0
has_power_backup           0
has_gym                    0
has_pool                   0
has_club                   0
has_lift                   0
has_servant_room           0
is_gated                   0
ov_park                    0
ov_pool                    0
ov_club                    0
ov_main_road               0
ov_sea                     0
ov_others                  0
dist_to_cyber_city         0
dist_to_golf_road          0
dist_to_airport            0
dist_to_manesa

In [5]:
corr = fs.corr(numeric_only=True)

fig = px.imshow(
    corr,
    text_auto=True,          
    color_continuous_scale='RdBu_r',  
    title="Correlation Heatmap"
)

fig.update_layout(
    xaxis_title="Features",
    yaxis_title="Features",
    width=800,
    height=700
)

In [6]:
corr['price_in_cr'].sort_values(ascending= False)

price_in_cr           1.000000
area                  0.684736
bathroom              0.483537
bedRoom               0.430463
covered_parking       0.375217
total_parking         0.333063
dist_to_manesar       0.251471
balcony               0.186114
total_floor           0.130412
has_ac                0.110659
has_servant_room      0.083342
open_parking          0.082718
ov_pool               0.039229
ov_club               0.031840
has_pool              0.013350
has_power_backup      0.010802
ov_park               0.003564
ov_others            -0.005324
has_club             -0.006516
is_corner            -0.010723
ov_sea               -0.012135
has_gym              -0.013968
ov_main_road         -0.020319
is_gated             -0.034987
has_lift             -0.058254
dist_to_airport      -0.279059
dist_to_cyber_city   -0.294253
dist_to_golf_road    -0.322940
Name: price_in_cr, dtype: float64

In [7]:
df = fs.copy()

## Feature ranking

This section was rebuilt. The previous version had three defects, all of which
flattered or distorted the result:

1. **The target was log-transformed twice** — once when separating X and y, then
   again after the split. Every importance below was ranked against
   `log(log(price))`. Monotonic, so the ordering was not nonsense, but the top of
   the price distribution was compressed hard and the numbers did not mean what
   they were labelled.
2. **The ranking split was not the model's split.** It took its own 80/20
   unstratified split of the preprocessed table, while the model uses a
   stratified 60/20/20 on the *de-duplicated* table. Different partitions — so
   rows that end up in the final test set helped choose the features. Mild, but
   it is selection leakage by the same standard applied everywhere else here.
3. **RFE with default settings** silently kept half the features, and its rank
   was inverted into a score without that being stated.

The corrected derivation lives in `scripts/rank_features.py` and is imported
below rather than reimplemented, so this notebook cannot drift from the artifact
it produces. It reproduces the production selection and split positionally, then
ranks on the train+validation rows only — the test set is never touched.

In [ ]:
import sys
sys.path.insert(0, '../..')

from scripts.rank_features import compute_ranking
from src.feature_selection.feature_selection import SELECTED_FEATURES

scores = compute_ranking()
scores.round(4)

### Three signals, because each is biased differently

**Random forest impurity** favours high-cardinality features — more distinct
values means more places to split, regardless of whether those splits
generalise. **LightGBM gain** is the deployed model family, so its importances
describe the model actually being shipped; it also handles the 184 rows with
unresolved coordinates natively, which sklearn's booster cannot. **Permutation
importance** is measured on held-out rows and asks the question that matters:
how much worse does error get if this feature is scrambled?

Averaging them does not make any one correct. It makes it unlikely that a
feature lives or dies on a single method's bias.

In [ ]:
chosen = set(SELECTED_FEATURES)
top_n  = set(scores.head(len(SELECTED_FEATURES)).index)

print('ranked but not selected:', sorted(top_n - chosen))
print('selected but not top-N :', sorted(chosen - top_n))
print()
print(scores['final_score'].tail(14).round(5).to_string())

### Conclusion — 24 features

The ranking confirms the top of the list decisively: area dominates, then
distance to Golf Course Road, total floors, property type, bathrooms and
distance to Cyber City. Below roughly rank 20 every score sits at 0.0002 or
less, which is noise — the ordering down there is not meaningful, and that is
the same fact as accuracy being flat across top-15 / top-25 / all-34. Trees
already down-weight weak features, so selection did not buy accuracy.

What it bought was a smaller serving surface (every feature kept is a field the
app must collect and validate), a more readable attribution story, and the
forcing of the question below.

**Two deliberate departures from the raw ranking:**

- **`society` is excluded despite ranking 7th.** It is close to a lookup of
  price — it encodes builder, age, amenity level, micro-location and buyer
  segment at once. Measured: **10.65%** MAPE with true values, **14.29%** with
  the ~31%-accurate guess production could actually obtain, **11.18%** retrained
  without it. Only the third describes something shippable. A feature's value is
  its importance *conditional on the quality of the value available at
  prediction time*, and society scores brilliantly while being worth negative
  points in production. It stays in the preprocessed data because analytics and
  the recommender use it legitimately — neither predicts anything.
- **`floornum_category` and `ov_main_road` are kept over `has_servant_room`**,
  which ranks slightly higher. Their scores differ by less than 0.0001, so this
  is a coin-flip dressed as a decision; the two kept fields are collected by the
  app, which is the tiebreak. Worth stating plainly rather than pretending the
  ranking drove it.

These 24 are hardcoded as `SELECTED_FEATURES` so the pipeline skips this
ranking and just selects columns — the feature set is part of the contract with
the front end and must not change silently between runs. `tests/test_feature_schema.py`
fails if the data and that list ever disagree.

**De-duplication.** Narrowing to these 24 collapses **823 rows** (38,858 →
38,035) that were genuinely distinct in the wider table — they differed on
`society` or on a dropped view flag. Earlier de-duplication passes could not
catch them; the duplication is *created* by feature selection, so it has to be
handled here. If a pair straddles the split the model recalls instead of
predicting.